### PR TITLE
JAMES-2551 rabbitmq should load mailqueue view configuration

### PR DESCRIPTION
**Pull request by trantienduchn**
This is a patch for #1775 to helps rabbitmq load `mailqueue view configuration` from rabbitmq.properties.

My work is bounded in last two commits

### CODE REVIEW

File: backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/CassandraMailQueueViewConfiguration.java [hunk 1]
```diff
@@ -17,11 +17,13 @@
  * under the License.                                           *
  ****************************************************************/
 
-package org.apache.james.queue.rabbitmq.view.cassandra.configuration;
+package org.apache.james.backend.rabbitmq;
```
Comment (chibenwa) [inline backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/CassandraMailQueueViewConfiguration.java]: I DISAGREE with this move.

That have nothing to do with `backend` and this code is not generic but specific to the mail queue.

File: backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQConfiguration.java [hunk 2]
```diff
@@ -102,13 +107,16 @@ public final int hashCode() {
...
         private Optional<Integer> maxRetries;
         private Optional<Integer> minDelay;
 
-        private Builder(URI amqpUri, URI managementUri, ManagementCredentials managementCredentials) {
+        private Builder(URI amqpUri, URI managementUri, ManagementCredentials managementCredentials,
+                        CassandraMailQueueViewConfiguration cassandraMailQueueViewConfiguration) {
             this.amqpUri = amqpUri;
             this.managementUri = managementUri;
             this.managementCredentials = managementCredentials;
+            this.cassandraMailQueueViewConfiguration = cassandraMailQueueViewConfiguration;
```
Comment (chibenwa) [inline backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQConfiguration.java]: No! We should not include a specific conf in a generic one...

Please note that sharing the same file != handling the conf reading in the same place.

File: backends-common/rabbitmq/src/test/java/org/apache/james/backend/rabbitmq/CassandraMailQueueViewConfigurationTest.java [hunk 3]
```diff
@@ -17,7 +17,7 @@
  * under the License.                                           *
```
Comment (chibenwa) [inline backends-common/rabbitmq/src/test/java/org/apache/james/backend/rabbitmq/CassandraMailQueueViewConfigurationTest.java]: Idem do not move this

File: server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJmapTestRule.java [hunk 4]
```diff
@@ -19,7 +19,9 @@
```
Comment (chibenwa) [inline server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJmapTestRule.java]: Given my previous comments, I of course disagree with the purpose of this commit.

**Comment by aduprat** [comment]
There's a compilation error:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:testCompile (default-testCompile) on project james-server-queue-rabbitmq: Compilation failure
[ERROR] /james-project/server/queue/queue-rabbitmq/src/test/java/org/apache/james/queue/rabbitmq/RabbitMqMailQueueFactoryTest.java:[93,13] cannot find symbol
[ERROR]   symbol:   method threadCount(int)
[ERROR]   location: interface org.apache.james.util.concurrency.ConcurrentTestRunner.RequireOperation
```

**Comment by trantienduchn** [comment]
> There's a compilation error:

Caught is Merging concurrent test PR just happens. fixuped and squashed in one commit

**Comment by trantienduchn** [comment]
```
[fe502dbcbb4a9417d486fb9ba034061bfc21ade9] [ERROR] Failures: 
[fe502dbcbb4a9417d486fb9ba034061bfc21ade9] [ERROR]   CassandraRabbitMQJamesServerTest.userFromLdapShouldLoginViaImapProtocol:85 expected:<[tru]e> but was:<[fals]e>
[fe502dbcbb4a9417d486fb9ba034061bfc21ade9] [ERROR] Errors: 
[fe502dbcbb4a9417d486fb9ba034061bfc21ade9] [ERROR]   CassandraRabbitMQJamesServerTest.mailsShouldBeWellReceivedBeforeFirstUserConnectionWithLdap:91 » SMTPSending
[fe502dbcbb4a9417d486fb9ba034061bfc21ade9] [INFO] 
[fe502dbcbb4a9417d486fb9ba034061bfc21ade9] [ERROR] Tests run: 4, Failures: 1, Errors: 1, Skipped: 0
```

**Comment by trantienduchn** [comment]
```
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [INFO] -------------------------------------------------------------
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [ERROR] COMPILATION ERROR : 
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [INFO] -------------------------------------------------------------
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [ERROR] /james-project/server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJamesServerTest.java:[41,55] cannot find symbol
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   symbol: class AbstractJmapJamesServerTest
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [ERROR] /james-project/server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJmapJamesServerTest.java:[28,59] cannot find symbol
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   symbol: class AbstractJmapJamesServerTest
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [ERROR] /james-project/server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJamesServerTest.java:[61,5] method does not override or implement a method from a supertype
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [ERROR] /james-project/server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJamesServerTest.java:[63,102] cannot find symbol
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   symbol:   variable DOMAIN_LIST_CONFIGURATION_MODULE
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   location: class org.apache.james.CassandraRabbitMQJamesServerTest
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [ERROR] /james-project/server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJamesServerTest.java:[71,5] method does not override or implement a method from a supertype
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [ERROR] /james-project/server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJamesServerTest.java:[78,9] cannot find symbol
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   symbol:   variable server
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   location: class org.apache.james.CassandraRabbitMQJamesServerTest
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [ERROR] /james-project/server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJamesServerTest.java:[82,31] cannot find symbol
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   symbol:   variable JAMES_SERVER_HOST
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   location: class org.apache.james.CassandraRabbitMQJamesServerTest
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [ERROR] /james-project/server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJamesServerTest.java:[82,50] cannot find symbol
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   symbol:   variable server
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   location: class org.apache.james.CassandraRabbitMQJamesServerTest
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [ERROR] /james-project/server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJamesServerTest.java:[85,33] cannot find symbol
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   symbol:   variable server
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   location: class org.apache.james.CassandraRabbitMQJamesServerTest
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [ERROR] /james-project/server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJamesServerTest.java:[87,35] cannot find symbol
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   symbol:   variable JAMES_SERVER_HOST
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   location: class org.apache.james.CassandraRabbitMQJamesServerTest
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [ERROR] /james-project/server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJamesServerTest.java:[87,54] cannot find symbol
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   symbol:   variable server
[76b8a1e7a471127032298c6ec73631c1b8fdd979]   location: class org.apache.james.CassandraRabbitMQJamesServerTest
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [ERROR] /james-project/server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJmapJamesServerTest.java:[38,5] method does not override or implement a method from a supertype
[76b8a1e7a471127032298c6ec73631c1b8fdd979] [ERROR] /james-project/server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJmapJamesServerTest.java:[43,5] method does not override or implement a method from a supertype
```

**Comment by chibenwa** [comment]
```

[c887919f7291ec8e4adf1796e2df80d9ee5048cd] [ERROR] Errors: 
[c887919f7291ec8e4adf1796e2df80d9ee5048cd] [ERROR]   RabbitMqMailQueueFactoryTest.createQueueShouldReturnTheSameInstanceWhenParallelCreateSameQueueName:97 » Execution
```

Oups looks like unstable

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/configuration/CassandraMailQueueViewConfiguration.java [hunk 5]
```diff
@@ -67,6 +69,30 @@ public CassandraMailQueueViewConfiguration build() {
...
+    public static CassandraMailQueueViewConfiguration from(Configuration configuration) {
+        Integer bucketCount = configuration.getInteger(BUCKET_COUNT_PROPERTY, null);
+        Preconditions.checkState(bucketCount != null, "You need to specify the " +
+            BUCKET_COUNT_PROPERTY + " property as the number of nodes from your cassandra cluster");
+
+        Integer updateBrowseStartPace = configuration.getInteger(UPDATE_BROWSE_START_PACE_PROPERTY, null);
+        Preconditions.checkState(updateBrowseStartPace != null, "You need to specify the " +
+            UPDATE_BROWSE_START_PACE_PROPERTY + " property as the average number of en-queue times, then after that, updating cassandra browse start point");
+
+        Long sliceWindowInSecond = configuration.getLong(SLICE_WINDOW_PROPERTY, null);
```
Comment (chibenwa) [inline server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/configuration/CassandraMailQueueViewConfiguration.java]: No this is not fiendly enough.

IMO we need to handle units ( and not a second count). Please handle this as a String and have a look to **TimeConverter** class

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/configuration/CassandraMailQueueViewConfiguration.java [hunk 6]
```diff
@@ -67,6 +69,30 @@ public CassandraMailQueueViewConfiguration build() {
         return bucketCount -> updateBrowseStartPace -> sliceWindow -> new Builder.ReadyToBuild(bucketCount, updateBrowseStartPace, sliceWindow);
     }
 
+    public static CassandraMailQueueViewConfiguration from(Configuration configuration) {
+        Integer bucketCount = configuration.getInteger(BUCKET_COUNT_PROPERTY, null);
+        Preconditions.checkState(bucketCount != null, "You need to specify the " +
+            BUCKET_COUNT_PROPERTY + " property as the number of nodes from your cassandra cluster");
```
Comment (chibenwa) [inline server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/configuration/CassandraMailQueueViewConfiguration.java]: No, not really. The buckets won't magically distribute accross the cluster evenly but rather randomly.

Thus one admin may wish to have more buckets in order to be sure of the repartition.

Exemple: With 3 node and 3 buckets I have a chance out of 27 to have all the buckets on the same node.

With one node and 9 buckets this is almost impossible.

I propose:

```
"You need to specify the " +
            BUCKET_COUNT_PROPERTY + " property. The higher the bucket count is, the better the data will be spread in the cluster. You might want the bucket count to exceed your Cassandra node count."
```

**Comment by trantienduchn** [comment]
```
[c887919f7291ec8e4adf1796e2df80d9ee5048cd] [ERROR] createQueueShouldReturnTheSameInstanceWhenParallelCreateSameQueueName  Time elapsed: 6.613 s  <<< ERROR!
[c887919f7291ec8e4adf1796e2df80d9ee5048cd] java.util.concurrent.ExecutionException: java.lang.RuntimeException: feign.FeignException: status 500 reading Api#listQueues()
[c887919f7291ec8e4adf1796e2df80d9ee5048cd] 	at org.apache.james.queue.rabbitmq.RabbitMqMailQueueFactoryTest.createQueueShouldReturnTheSameInstanceWhenParallelCreateSameQueueName(RabbitMqMailQueueFactoryTest.java:97)
[c887919f7291ec8e4adf1796e2df80d9ee5048cd] Caused by: java.lang.RuntimeException: feign.FeignException: status 500 reading Api#listQueues()
[c887919f7291ec8e4adf1796e2df80d9ee5048cd] Caused by: feign.FeignException: status 500 reading Api#listQueues()
[c887919f7291ec8e4adf1796e2df80d9ee5048cd] 	at 
```
came from rabbitmq management. test this please

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/configuration/CassandraMailQueueViewConfiguration.java [hunk 7]
```diff
@@ -67,6 +69,30 @@ public CassandraMailQueueViewConfiguration build() {
...
+
+        Integer updateBrowseStartPace = configuration.getInteger(UPDATE_BROWSE_START_PACE_PROPERTY, null);
+        Preconditions.checkState(updateBrowseStartPace != null, "You need to specify the " +
+            UPDATE_BROWSE_START_PACE_PROPERTY + " property as the average number of en-queue times, then after that, updating cassandra browse start point");
```
Comment (chibenwa) [inline server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/configuration/CassandraMailQueueViewConfiguration.java]: Well idem the message is not clear:

```
"You need to specify the " +
            UPDATE_BROWSE_START_PACE_PROPERTY + " property. The mailQueue will use this number to randomly update the browsing start point of the mail queue. The higher that number is, the more the start point will be up to date but the more unnecessary work will be performed.");
```

File: server/container/guice/rabbitmq/src/main/java/org/apache/james/modules/rabbitmq/RabbitMQModule.java [hunk 8]
```diff
@@ -105,9 +98,20 @@ protected void configure() {
...
     @Singleton
     private RabbitMQConfiguration getMailQueueConfiguration(PropertiesProvider propertiesProvider) throws ConfigurationException {
+        Configuration configuration = loadConfigurationFiles(propertiesProvider);
+        return RabbitMQConfiguration.from(configuration);
+    }
+
+    @Provides
+    @Singleton
+    private CassandraMailQueueViewConfiguration getMailQueueViewConfiguration(PropertiesProvider propertiesProvider) throws ConfigurationException {
+        Configuration configuration = loadConfigurationFiles(propertiesProvider);
```
Comment (chibenwa) [inline server/container/guice/rabbitmq/src/main/java/org/apache/james/modules/rabbitmq/RabbitMQModule.java]: Cqn you add your entries to the sample configuration file (in test resources ?)

**Comment by trantienduchn** [comment]
```
org.apache.james.queue.rabbitmq.RabbitMqMailQueueFactoryTest
[c887919f7291ec8e4adf1796e2df80d9ee5048cd] [ERROR] createQueueShouldReturnTheSameInstanceWhenParallelCreateSameQueueName  Time elapsed: 6.514 s  <<< ERROR!
[c887919f7291ec8e4adf1796e2df80d9ee5048cd] java.util.concurrent.ExecutionException: java.lang.RuntimeException: feign.FeignException: status 500 reading Api#listQueues()
[c887919f7291ec8e4adf1796e2df80d9ee5048cd] 	at org.apache.james.queue.rabbitmq.RabbitMqMailQueueFactoryTest.createQueueShouldReturnTheSameInstanceWhenParallelCreateSameQueueName(RabbitMqMailQueueFactoryTest.java:97)
[c887919f7291ec8e4adf1796e2df80d9ee5048cd] Caused by: java.lang.RuntimeException: feign.FeignException: status 500 reading Api#listQueues()
[c887919f7291ec8e4adf1796e2df80d9ee5048cd] Caused by: feign.FeignException: status 500 reading Api#listQueues()
[c887919f7291ec8e4adf1796e2df80d9ee5048cd] 	at org.apache.james.queue.rabbitmq.RabbitMqMailQueueFactoryTest.lambda$createQueueShouldReturnTheSameInstanceWhenParallelCreateSameQueueName$0(RabbitMqMailQueueFactoryTest.java:94)
```

hmm...

**Comment by trantienduchn** [comment]
```
[7cfd4eac3f347ab01ce44be48d71df102861ce3e] [ERROR] Errors: 
[7cfd4eac3f347ab01ce44be48d71df102861ce3e] [ERROR]   RabbitMQClusterTest$ClusterSharing.queuesShouldBeDeclarableOnAnotherNode » Completion
[7cfd4eac3f347ab01ce44be48d71df102861ce3e] [INFO] 
[7cfd4eac3f347ab01ce44be48d71df102861ce3e] [ERROR] Tests run: 39, Failures: 0, Errors: 1, Skipped: 1
```

test this please

**Comment by trantienduchn** [comment]
> Could you add tests for configuration parsing in the following case

Here you are :)

**Comment by chibenwa** [comment]
test this please

**Comment by trantienduchn** [comment]
```
[dabeb5d8ecad1691e06962958c971aa76de3c41e] [INFO] Tests run: 38, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 121.925 s - in org.apache.james.jmap.cassandra.CassandraGetMailboxesMethodTest
[dabeb5d8ecad1691e06962958c971aa76de3c41e] [INFO] 
[dabeb5d8ecad1691e06962958c971aa76de3c41e] [INFO] Results:
[dabeb5d8ecad1691e06962958c971aa76de3c41e] [INFO] 
[dabeb5d8ecad1691e06962958c971aa76de3c41e] [ERROR] Errors: 
[dabeb5d8ecad1691e06962958c971aa76de3c41e] [ERROR]   CassandraMailboxSharingTest » Runtime javax.management.InstanceAlreadyExistsEx...
[dabeb5d8ecad1691e06962958c971aa76de3c41e] [ERROR]   CassandraMailboxSharingTest » Runtime org.apache.commons.configuration.Configu...
[dabeb5d8ecad1691e06962958c971aa76de3c41e] [ERROR]   CassandraMailboxSharingTest » ContainerLaunch Container startup failed
[dabeb5d8ecad1691e06962958c971aa76de3c41e] [ERROR]   CassandraMailboxSharingTest » Provision Unable to provision, see the following...
[dabeb5d8ecad1691e06962958c971aa76de3c41e] [INFO] 
[dabeb5d8ecad1691e06962958c971aa76de3c41e] [ERROR] Tests run: 536, Failures: 0, Errors: 4, Skipped: 3
[dabeb5d8ecad1691e06962958c971aa76de3c41e] [INFO] 
```
test this please

**Comment by trantienduchn** [comment]
cherry-pick from #1775 (new rabbitmq product) to fasten test (387b400, bf973bd)

**Comment by chibenwa** [comment]
This seems done

**Pull request closed without merging**